Adds support to specify linux_fx_version for Function Apps

joakimhew opened the pull request:
This PR fixes [1937](https://github.com/terraform-providers/terraform-provider-azurerm/issues/1937) and allows users to specify the `linux_fx_version` for azure function apps ([resource_arm_function_app.go](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/resource_arm_function_app.go))

For consistency, I think `linux_fx_version` is a better option than the alternative mentioned in the issue, as `linux_fx_version` is already used in App Services.

joakimhew commented:
@katbyte Thank you for the review! Should've thought of that of course haha. I'll correct it later and update my PR 👍

joakimhew commented:
@katbyte I've just realised that we also need to allow the user to set the `kind` field for the function. To enable container based function this need to be set to something like `functionapp,linux,container` 

I'll add `kind` to the schema and tests for it and update my PR

katbyte commented on azurerm/resource_arm_function_app.go:
```diff
@@ -37,6 +37,12 @@ func resourceArmFunctionApp() *schema.Resource {
 
+			"kind": {
+				Type:     schema.TypeString,
+				Optional: true,
```
I don't see you using the property anywhere, just reading it into state. Is that an oversight or should just maybe just be `Optional`?

katbyte replied:
It is also not set in any of your test configs or documented

katbyte replied:
We can just mark it as `optional` (i believe that is what i wanted to put in my above comment 😅), document it and then continue to use it in the acc test

joakimhew replied:
Ahhh, okay I thought you meant the `Optional` in the schema. But you mean marking it as `Optional` in the documentation too 😅? Sure I'll get to it later :) Thanks a lot for the help @katbyte

joakimhew replied:
Hi again @katbyte, hope you're good!

I'm scratching my head around the `kind` attribute. Since changing this attribute has no affect and will always be computed, wouldn't it confuse the user to see this value in the schema?

I currently see two possible solutions for it.
1. Use `Optional: true` and `Computed: true` and then specify in the docs that this is computed and changing it will have no effect.
2. Use `UnsafeSetFieldRaw` to be able to set the attribute to the schema in `resourceArmFunctionAppRead` so that we can confirm the `kind` in our acceptance tests.

What do you think?

katbyte commented on azurerm/resource_arm_function_app_test.go:
```diff
@@ -1024,27 +1112,85 @@ resource "azurerm_storage_account" "test" {
     size = "S1"
+  }
+	
+	properties {
```
Could we use spaces instead of tabs for indentation here?

joakimhew replied:
Should've done what I just did from the beginning... 😄
`cmd + shift + p` -> `Convert Indentation to Spaces`

joakimhew commented:
@katbyte Once we've resolved the last thing, I'd like to fix the author of my commits as they are currently not set to my GitHub account. This would rewrite history, is that okay?

katbyte commented:
@joakimhew, that is totally fine 🙂 just @me when your ready for a final review and merge!

joakimhew commented:
@katbyte Just set `kind` in schema to only computed and added it as an attribute reference to the docs. I've also changed the author of my commits so everything should be resolved now 🎊 Thanks again!

katbyte commented on website/docs/r/function_app.html.markdown:
```diff
@@ -157,6 +157,8 @@ The following attributes are exported:
 
 * `site_credential` - A `site_credential` block as defined below, which contains the site-level credentials used to publish to this App Service.
 
+* `kind` - The Function App kind - such as `functionapp,linux,container`
```
Could we separate these values:
```suggestion
* `kind` - The Function App kind - such as `functionapp`, `linux`, `container`
```

katbyte commented:
tests:
```
[12:30:02] kt@snowbook:~/hashi/..3../terraform-providers/terraform-provider-azurerm▸features/function_app_linux_fx_version$ testazure TestAccAzureRMFunctionApp
==> Fixing source code with gofmt...
# This logic should match the search logic in scripts/gofmtcheck.sh
gofmt -s -w `find . -name '*.go' | grep -v vendor`
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./azurerm -v -test.run=TestAccAzureRMFunctionApp -timeout 180m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAzureRMFunctionApp_basic
=== PAUSE TestAccAzureRMFunctionApp_basic
=== RUN   TestAccAzureRMFunctionApp_requiresImport
=== PAUSE TestAccAzureRMFunctionApp_requiresImport
=== RUN   TestAccAzureRMFunctionApp_tags
=== PAUSE TestAccAzureRMFunctionApp_tags
=== RUN   TestAccAzureRMFunctionApp_tagsUpdate
=== PAUSE TestAccAzureRMFunctionApp_tagsUpdate
=== RUN   TestAccAzureRMFunctionApp_appSettings
=== PAUSE TestAccAzureRMFunctionApp_appSettings
=== RUN   TestAccAzureRMFunctionApp_siteConfig
=== PAUSE TestAccAzureRMFunctionApp_siteConfig
=== RUN   TestAccAzureRMFunctionApp_linuxFxVersion
=== PAUSE TestAccAzureRMFunctionApp_linuxFxVersion
=== RUN   TestAccAzureRMFunctionApp_connectionStrings
=== PAUSE TestAccAzureRMFunctionApp_connectionStrings
=== RUN   TestAccAzureRMFunctionApp_siteConfigMulti
=== PAUSE TestAccAzureRMFunctionApp_siteConfigMulti
=== RUN   TestAccAzureRMFunctionApp_updateVersion
=== PAUSE TestAccAzureRMFunctionApp_updateVersion
=== RUN   TestAccAzureRMFunctionApp_3264bit
=== PAUSE TestAccAzureRMFunctionApp_3264bit
=== RUN   TestAccAzureRMFunctionApp_httpsOnly
=== PAUSE TestAccAzureRMFunctionApp_httpsOnly
=== RUN   TestAccAzureRMFunctionApp_consumptionPlan
=== PAUSE TestAccAzureRMFunctionApp_consumptionPlan
=== RUN   TestAccAzureRMFunctionApp_consumptionPlanUppercaseName
=== PAUSE TestAccAzureRMFunctionApp_consumptionPlanUppercaseName
=== RUN   TestAccAzureRMFunctionApp_createIdentity
=== PAUSE TestAccAzureRMFunctionApp_createIdentity
=== RUN   TestAccAzureRMFunctionApp_updateIdentity
=== PAUSE TestAccAzureRMFunctionApp_updateIdentity
=== RUN   TestAccAzureRMFunctionApp_loggingDisabled
=== PAUSE TestAccAzureRMFunctionApp_loggingDisabled
=== RUN   TestAccAzureRMFunctionApp_updateLogging
=== PAUSE TestAccAzureRMFunctionApp_updateLogging
=== CONT  TestAccAzureRMFunctionApp_basic
=== CONT  TestAccAzureRMFunctionApp_3264bit
=== CONT  TestAccAzureRMFunctionApp_siteConfig
=== CONT  TestAccAzureRMFunctionApp_updateVersion
=== CONT  TestAccAzureRMFunctionApp_siteConfigMulti
=== CONT  TestAccAzureRMFunctionApp_updateLogging
=== CONT  TestAccAzureRMFunctionApp_loggingDisabled
=== CONT  TestAccAzureRMFunctionApp_updateIdentity
=== CONT  TestAccAzureRMFunctionApp_createIdentity
--- PASS: TestAccAzureRMFunctionApp_basic (153.15s)
--- PASS: TestAccAzureRMFunctionApp_loggingDisabled (154.82s)
=== CONT  TestAccAzureRMFunctionApp_consumptionPlanUppercaseName
--- PASS: TestAccAzureRMFunctionApp_siteConfig (156.64s)
=== CONT  TestAccAzureRMFunctionApp_consumptionPlan
--- PASS: TestAccAzureRMFunctionApp_updateVersion (176.83s)
=== CONT  TestAccAzureRMFunctionApp_httpsOnly
--- PASS: TestAccAzureRMFunctionApp_updateIdentity (185.98s)
=== CONT  TestAccAzureRMFunctionApp_tags
--- PASS: TestAccAzureRMFunctionApp_3264bit (186.47s)
=== CONT  TestAccAzureRMFunctionApp_tagsUpdate
--- PASS: TestAccAzureRMFunctionApp_updateLogging (216.43s)
=== CONT  TestAccAzureRMFunctionApp_requiresImport
--- PASS: TestAccAzureRMFunctionApp_createIdentity (139.54s)
=== CONT  TestAccAzureRMFunctionApp_appSettings
--- PASS: TestAccAzureRMFunctionApp_httpsOnly (148.47s)
=== CONT  TestAccAzureRMFunctionApp_connectionStrings
--- PASS: TestAccAzureRMFunctionApp_consumptionPlan (176.59s)
=== CONT  TestAccAzureRMFunctionApp_linuxFxVersion
--- PASS: TestAccAzureRMFunctionApp_siteConfigMulti (339.27s)
--- PASS: TestAccAzureRMFunctionApp_tags (156.16s)
--- PASS: TestAccAzureRMFunctionApp_consumptionPlanUppercaseName (188.42s)
--- PASS: TestAccAzureRMFunctionApp_tagsUpdate (172.81s)
--- PASS: TestAccAzureRMFunctionApp_requiresImport (158.40s)
--- PASS: TestAccAzureRMFunctionApp_connectionStrings (151.61s)
--- PASS: TestAccAzureRMFunctionApp_linuxFxVersion (162.18s)
--- PASS: TestAccAzureRMFunctionApp_appSettings (219.76s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm	512.490s
[12:40:27] kt@snowbook:~/hashi/..3../terraform-providers/terraform-provider-azurerm▸features/function_app_linux_fx_version$
```
